File: shared/BuddyCore/Sources/BuddyCore/Domain/HostElection.swift
import Foundation

/// Deterministic host election for a 2-peer session. Both peers run this
/// locally on the same `(a, b)` tuple and agree on the host without
/// negotiating.
///
/// Algorithm:
///   1. The peer with the lexicographically SMALLER `id.uuidString` wins.
///   2. Tie-break (impossible with v4 UUIDs but kept for theoretical
///      completeness): iOS wins over Android.
public enum HostElection {

    public static func host(between a: Peer, _ b: Peer) -> Peer {
        let aStr = a.id.uuidString
        let bStr = b.id.uuidString
        if aStr < bStr { return a }
        if bStr < aStr { return b }
        // Same UUID — extremely unlikely. Tie-break by platform.
        switch (a.platform, b.platform) {
        case (.ios, .android): return a
        case (.android, .ios): return b
        default:                return a
        }
    }

    /// Convenience: which of the two is the guest.
    public static func guest(between a: Peer, _ b: Peer) -> Peer {
        host(between: a, b).id == a.id ? b : a
    }
}

File: apps/BuddyPlay/android/core/src/main/java/com/americangroupllc/buddyplay/core/domain/HostElection.kt
package com.americangroupllc.buddyplay.core.domain

import com.americangroupllc.buddyplay.core.models.Peer

/**
 * Deterministic host election for a 2-peer session. Both peers run this
 * locally on the same `(a, b)` tuple and agree on the host without
 * negotiating.
 *
 * Algorithm:
 *   1. The peer with the lexicographically SMALLER `id` wins.
 *   2. Tie-break (impossible with v4 UUIDs but kept for theoretical
 *      completeness): iOS wins over Android.
 */
object HostElection {

    fun host(a: Peer, b: Peer): Peer {
        val cmp = a.id.compareTo(b.id)
        if (cmp < 0) return a
        if (cmp > 0) return b
        // Same id — tie-break on platform.
        return when {
            a.platform == Peer.Platform.IOS && b.platform == Peer.Platform.ANDROID -> a
            b.platform == Peer.Platform.IOS && a.platform == Peer.Platform.ANDROID -> b
            else -> a
        }
    }

    fun guest(a: Peer, b: Peer): Peer = if (host(a, b).id == a.id) b else a
}
